CRM/Core/BAO/Attachment.php

```diff
@@ -0,0 +1,282 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+
+/**
+ * Business objects for managing Attachment.
+ */
+class CRM_Core_BAO_Attachment extends CRM_Core_DAO {
+
+  /**
+   * Create a dedupe exception record.
+   *
+   * @param array $params
+   *
+   * @return \CRM_Dedupe_BAO_Exception
+   */
+  public static function create($params) {
+    $config = CRM_Core_Config::singleton();
+    list($id, $file, $entityFile, $name, $content, $moveFile, $isTrusted, $returnContent) = self::parseParams($params);
+
+    $fileDao = new CRM_Core_BAO_File();
+    $entityFileDao = new CRM_Core_DAO_EntityFile();
+
+    if ($id) {
+      $file['id'] = $fileDao->id = $id;
+
+      if (!$fileDao->find(TRUE)) {
+        throw new CRM_Core_Exception("Invalid ID");
+      }
+
+      $entityFileDao->file_id = $id;
+      if (!$entityFileDao->find(TRUE)) {
+        throw new CRM_Core_Exception("Cannot modify orphaned file");
+      }
+    }
+
+    if (!$id && !is_string($content) && !is_string($moveFile)) {
+      throw new CRM_Core_Exception("Mandatory key(s) missing from params array: 'id' or 'content' or 'options.move-file'");
+    }
+    if (!$isTrusted && $moveFile) {
+      throw new CRM_Core_Exception("options.move-file is only supported on secure calls");
+    }
+    if (is_string($content) && is_string($moveFile)) {
+      throw new CRM_Core_Exception("'content' and 'options.move-file' are mutually exclusive");
+    }
+    if ($id && !$isTrusted && isset($file['upload_date']) && $file['upload_date'] != CRM_Utils_Date::isoToMysql($fileDao->upload_date)) {
+      throw new CRM_Core_Exception("Cannot modify upload_date" . var_export([$file['upload_date'], $fileDao->upload_date, CRM_Utils_Date::isoToMysql($fileDao->upload_date)], TRUE));
+    }
+    if ($id && $name && $name != CRM_Utils_File::cleanFileName($fileDao->uri)) {
+      throw new CRM_Core_Exception("Cannot modify name");
+    }
```

Comment:
A lot of this could probably be handled by the `required_if` fieldSpec properties. IIRC you were the one who wrote that bit to handle expressions using smarty.

Reply:
Oh sorry nevermind this is a BAO function.